api/src/services/graphql/utils/replace-m2a.ts

```diff
@@ -0,0 +1,81 @@
+import type { Filter, NestedDeepQuery, SchemaOverview } from '@directus/types';
+
+export function filterReplaceM2A(filter_arg: Filter, collection: string, schema: SchemaOverview): any {
+	const filter: any = filter_arg;
+
+	for (const key in filter) {
+		const o2m_relation = schema.relations.find(
+			(relation) => relation.related_collection === collection && relation.meta?.one_field === key,
+		);
+
+		const m2o_relation = schema.relations.find(
+			(relation) => relation.collection === collection && relation.field === key,
+		);
+
+		const a2o_relation = schema.relations.find(
+			(relation) =>
+				relation.collection === collection &&
+				key.startsWith(relation.field + '__') &&
+				relation.meta?.one_allowed_collections?.some((collection) => collection === key.split('__')[1]),
+		);
+
+		if (o2m_relation) {
+			filter[key] = filterReplaceM2A(filter[key], o2m_relation.collection!, schema);
+		} else if (m2o_relation) {
+			filter[key] = filterReplaceM2A(filter[key], m2o_relation.related_collection!, schema);
+		} else if (a2o_relation) {
+			const new_key = key.replace('__', ':');
+			filter[new_key] = filterReplaceM2A(filter[key], key.split('__')[1]!, schema);
+			delete filter[key];
+		} else if (Array.isArray(filter[key])) {
+			filter[key] = filter[key].map((item) => filterReplaceM2A(item, collection, schema));
+		} else if (typeof filter[key] === 'object') {
+			filter[key] = filterReplaceM2A(filter[key], collection, schema);
+		}
+	}
+
+	return filter;
+}
+
+export function filterReplaceM2ADeep(
+	deep_arg: NestedDeepQuery | null | undefined,
+	collection: string,
+	schema: SchemaOverview,
+) {
+	if (!deep_arg) return deep_arg;
```

Comment:
For the sake of consistency should we update `filterReplaceM2A` to do the same as `filterReplaceM2ADeep` and return for `null`/`undefined` filter? Currently we require `filter` for `filterReplaceM2A` but not `deep` for `filterReplaceM2ADeep`